File: src/main/resources/static/ts/GameObjects/Room.ts
import GameObject = require('./GameObject');
import Solid = require('./Solid');
import Rect = require('../Rect');

export = class Room extends GameObject {
    private readonly width: number;
    ceiling: Solid;
    wallLeft: Solid;
    wallRight: Solid;
    floor: Solid;

    constructor(parent: GameObject | null,
                frame: Rect,
                width: number,
                damageVelocityThreshold: number,
                damageVelocityMultiplier: number
    ) {
        super(parent, frame);

        this.width = width;
        this.ceiling = new Solid(this, new Rect(
            0,
            -this.frame.height / 2 + this.width / 2,
            this.frame.width,
            this.width), damageVelocityThreshold, damageVelocityMultiplier);
        this.wallLeft = new Solid(this, new Rect(
            -this.frame.width / 2 + this.width / 2,
            0,
            this.width,
            this.frame.height - this.width * 2), damageVelocityThreshold, damageVelocityMultiplier);
        this.wallRight = new Solid(this, new Rect(
            this.frame.width / 2 - this.width / 2,
            0,
            this.width,
            this.frame.height - this.width * 2), damageVelocityThreshold, damageVelocityMultiplier);
        this.floor = new Solid(this, new Rect(
            0,
            this.frame.height / 2 - this.width / 2,
            this.frame.width,
            this.width), damageVelocityThreshold, damageVelocityMultiplier);
    };
}

File: src/main/kotlin/me/demoniacdeath/onlineplatformer/Room.kt
package me.demoniacdeath.onlineplatformer

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonIgnoreProperties

@JsonIgnoreProperties("children")
class Room(rect: Rect,
           val width: Double,
           @Suppress("CanBeParameter") val damageVelocityThreshold: Double,
           @Suppress("CanBeParameter") val damageVelocityMultiplier: Double
) : GameObject(rect) {
    override val type = "Room"
    @JsonIgnore
    val floor: Solid = Solid(Rect(
            .0,
            frame.size.height / 2 - width / 2,
            frame.size.width,
            width
    ), damageVelocityThreshold, damageVelocityMultiplier)
    @JsonIgnore
    val wallLeft: Solid = Solid(Rect(
            -frame.size.width / 2 + width / 2,
            .0,
            width,
            frame.size.height - width * 2
    ), damageVelocityThreshold, damageVelocityMultiplier)
    @JsonIgnore
    val wallRight: Solid = Solid(Rect(
            frame.size.width / 2 - width / 2,
            .0,
            width,
            frame.size.height - width * 2
    ), damageVelocityThreshold, damageVelocityMultiplier)
    @JsonIgnore
    val ceiling: Solid = Solid(Rect(
            .0,
            -frame.size.height / 2 + width / 2,
            frame.size.width,
            width
    ), damageVelocityThreshold, damageVelocityMultiplier)

    init {
        addChild(floor)
        addChild(wallLeft)
        addChild(wallRight)
        addChild(ceiling)
    }
}
